### PR TITLE
seed: implement Copier for seed20

### PR DESCRIPTION
**Pull request by andrewphelpsj**
This PR adds a `Copier` interface that seeds can implement to copy themselves to a given seed partition. Additionally, an implementation for seed20 is done here. This will be used as a part of installing seeds during UC installation via the snapd API.

### CODE REVIEW

File: seed/seed20.go [hunk 1]
```diff
@@ -82,6 +84,82 @@ type seed20 struct {
 	essentialSnapsNum int
 }
 
+// Copy implement Copier interface. Copies the seed to the given root, where
+// root is treated as the root of a seed partition.
+func (s *seed20) Copy(root string) (err error) {
+	// TODO: right now this method requires you to call LoadAssertions and
+	// LoadMeta first. is that okay? should Copier embed the Seed interface to
+	// make that easier? i could call them in here, but then i'd need to take in
+	// a db as a parameter.
```
Comment (pedronis) [inline seed/seed20.go:93]: this is fine I think but the doc comment on the Copier interface should explain this

File: seed/seed.go [hunk 2]
```diff
@@ -205,6 +205,14 @@ type PreseedCapable interface {
 	LoadPreseedAssertion() (*asserts.Preseed, error)
 }
 
+// Copier can be implemented by a seed that supports copying itself to a given
+// destination.
+type Copier interface {
+	// Copy copies the seed to the given seed partition. The root parameter
+	// is treated as the root of the seed partition.
+	Copy(root string) error
```
Comment (pedronis) [inline seed/seed.go:213]: I think this should be UC20+  and take seedDir and label like Open below, label could be "" in which case the same label is reused

**Review by pedronis — COMMENTED** [review]
thank you, did a first pass, couple small comments

File: seed/seed.go (hunk 2 above)
Comment (andrewphelpsj) [reply seed/seed.go:213]: Makes sense, will make that change.

File: seed/seed20.go (hunk 1 above)
Comment (andrewphelpsj) [reply seed/seed20.go:93]: In that case, should `Copier` embed `Seed`? Since `Copier` will requirer methods from `Seed`?

Comment (andrewphelpsj) [reply seed/seed20.go]: Comment is updated.

Comment (pedronis) [reply seed/seed20.go]: yes, including Seed is fine, if you look other interfaces do that too

File: seed/seed20.go [hunk 3]
```diff
@@ -82,6 +84,78 @@ type seed20 struct {
...
+	// copy the asserted snaps that the seed needs
+	for _, sn := range s.snaps {
+		// unasserted snaps are already copied above, skip them
+		if sn.ID() == "" {
+			continue
+		}
+
+		destPath := filepath.Join(destSeedDir, "snaps", filepath.Base(sn.Path))
+
+		if err := os.MkdirAll(filepath.Dir(destPath), 0755); err != nil {
```
Comment (pedronis) [inline seed/seed20.go:147]: we might want to try to do this only once

File: seed/seed.go [hunk 4]
```diff
@@ -205,6 +205,18 @@ type PreseedCapable interface {
...
 
+// Copier can be implemented by a seed that supports copying itself to a given
+// destination.
+type Copier interface {
+	Seed
+	// Copy copies the seed to the given seedDir with the label provided. If
+	// label is empty, then the label of the seed that implements Copier is
+	// used. This interface only makes sense to implement for UC20+ seeds. Copy
+	// requires you to call the methods LoadAssertions and LoadMeta first. Only
+	// the snaps for the mode that was passed to LoadMeta will be copied.
```
Comment (pedronis) [inline seed/seed.go:216]: I missed the last bit, that doesn't make sense, we need to copy the snaps for all the modes

**Review by pedronis — CHANGES_REQUESTED** [review]

File: seed/seed20.go (hunk 3 above)
Comment (andrewphelpsj) [reply seed/seed20.go:147]: I'm not sure I follow. `destPath` changes each loop, so we are only doing this once, I think?

File: seed/seed.go (hunk 4 above)
Comment (andrewphelpsj) [reply seed/seed.go:216]: In that case, I can move the call to LoadMeta into Copy. Although, that will require the caller to call LoadMeta again, if they happened to call it previsouly with a different mode.

File: seed/seed20.go (hunk 3 above)
Comment (andrewphelpsj) [reply seed/seed20.go:147]: Oh no, I can't read. I see.

File: seed/seed.go (hunk 4 above)
Comment (andrewphelpsj) [reply seed/seed.go]: I could check to see if the mode was changed from not-all to all, and then re-call LoadMeta at the end of the copy to put things back in place like they were before the method was called. I'm not sure if that is needed, though.

Comment (andrewphelpsj) [reply seed/seed.go]: That wouldn't preserve the effects of a custom `SnapHandler`, so probably not a good idea.

File: seed/seed20.go [hunk 5]
```diff
@@ -135,20 +135,26 @@ func (s *seed20) Copy(seedDir string, label string) (err error) {
 		return err
 	}
```
Comment (pedronis) [inline seed/seed20.go:137]: wouldn't be a bad idea for this to take a timings.Measurer and measure the time spent doing the copying after the meta loading

**Review by pedronis — APPROVED** [review]
thank you, remark about producing timings

File: seed/seed20_test.go [hunk 6]
```diff
@@ -3507,3 +3508,208 @@ func (s *seed20Suite) TestPreseedCapableSeedAlternateAuthority(c *C) {
...
+	copier, ok := seed20.(seed.Copier)
+	c.Assert(ok, Equals, true)
+
+	// remove a snap from the original seed to make the copy fail
+	err = os.Remove(filepath.Join(s.SeedDir, "snaps", "snapd_1.snap"))
+	c.Assert(err, IsNil)
+
+	destSeedDir := c.MkDir()
+	err = copier.Copy(destSeedDir, label, s.perfTimings)
+	c.Check(err, NotNil)
```
Comment (alfonsosanchezbeato) [inline seed/seed20_test.go:3656]: I think it would be better to check more specifically the error message.

File: seed/seed20_test.go [hunk 7]
```diff
@@ -3507,3 +3508,208 @@ func (s *seed20Suite) TestPreseedCapableSeedAlternateAuthority(c *C) {
...
+		}
+
+		c.Check(gotPath, testutil.FileEquals, testutil.FileContentRef(path))
+
+		return nil
+	})
+	c.Assert(err, IsNil)
+
+	gotCount := 0
+	err = filepath.WalkDir(expected, func(_ string, _ fs.DirEntry, err error) error {
```
Comment (alfonsosanchezbeato) [inline seed/seed20_test.go:3705]: Maybe I am getting this wrong, but shouldn't this run through `got` folder instead?

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks, just a couple of minor test related comments

File: seed/seed20_test.go (hunk 7 above)
Comment (andrewphelpsj) [reply seed/seed20_test.go:3705]: Good catch, thank you!

**Pull request merged**